gmeta/codegen/src/lib.rs

```diff
@@ -1,518 +1,503 @@
-    // Validation not required for `ident`.
-    ensure!(
-        signature.generics.params.is_empty(),
-        &signature.generics.params[0],
-        "Function shouldn't have generics"
-    );
-    ensure!(
-        signature.generics.where_clause.is_none(),
-        &signature.generics.where_clause,
-        "Function shouldn't have where clause"
-    );
-    // Validation not required for `parent_token`.
-    let validate_typed = |arg: &FnArg, state_name: bool| -> Result<(), TokenStream> {
-        match arg {
-            FnArg::Typed(pat_type) => {
-                ensure!(
-                    pat_type.attrs.is_empty(),
-                    &pat_type.attrs[0],
-                    "Arguments shouldn't have attributes"
-                );
-
-                ensure!(
-                    matches!(&*pat_type.ty, syn::Type::Path(_)),
-                    &*pat_type.ty,
-                    "Illegal type specification: should be path-based"
-                );
-
-                if let Pat::Ident(pat_ident) = &*pat_type.pat {
-                    ensure!(
-                        pat_ident.attrs.is_empty(),
-                        &pat_ident.attrs[0],
-                        "Pattern idents shouldn't have attributes"
-                    );
-                    ensure!(
-                        pat_ident.by_ref.is_none(),
-                        &pat_ident.by_ref,
-                        "Pattern idents shouldn't be thrown as ref"
-                    );
-                    ensure!(
-                        pat_ident.subpat.is_none(),
-                        &pat_ident.subpat.as_ref().expect("checked in cond").1,
-                        "Pattern idents shouldn't be thrown using subpattern"
-                    );
-
-                    if (pat_ident.ident == "state") != state_name {
-                        return Err(compile_error(
-                            &pat_ident.ident,
-                            if state_name {
-                                "Illegal name: should be `state`"
-                            } else {
-                                "Illegal name: shouldn't be `state`"
-                            },
-                        ));
-                    }
-                } else {
-                    return Err(compile_error(
-                        pat_type,
-                        "Illegal pattern type: use common `var: Type`",
-                    ));
-                }
+        let mut items = items.into_iter();
+        let two_first_items = (items.next(), items.next());
 
-                Ok(())
-            }
-            _ => Err(compile_error(arg, "Self arguments are restricted")),
+        if let (Some(first), Some(second)) = two_first_items {
+            (first, iter::once(second).chain(items))
+        } else {
+            error!(
+                module_span,
+                "module with #[metawasm] must contain at least 2 items"
+            );
         }
+    } else {
+        error!(
```

Comment:
move above, please. for macro fn bodies I prefer not to use too many logic branching